### PR TITLE
Add large_tuple rule

### LINKED ISSUE #1065: Rule request: large tuples
Tuples should have at most 3(?) members.

### PR DESCRIPTION
**Pull request by marcelofabri**
Fixes #1065 

~This currently triggers when tuples are used to represent closure parameters, such as:~

```swift
func setAndTestAllTypes(_ setter: (SwiftObject, Any?, String) -> Void,
                          getter: (SwiftObject, String) -> (Any?), object: SwiftObject)
```

~I'm not sure if we should handle this case or not.~

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/LargeTupleRule.swift [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+    private func parenthesesRanges(_ text: String) throws -> [NSRange] {
+        var stack = [Int]()
+        var balanced = true
+        var ranges = [NSRange]()
+
+        let nsText = text.bridge()
+        let fullRange = NSRange(location: 0, length: nsText.length)
+        let options = NSString.EnumerationOptions.byComposedCharacterSequences
+        nsText.enumerateSubstrings(in: fullRange, options: options) { (substring, range, _, stop) in
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/LargeTupleRule.swift]: `fatal error: enumerateSubstrings(in:options:using:) is not yet implemented` 🙃

**Comment by aamctustwo** [comment]
:+1: For ignoring closure declarations on this rule. I see closures as special types of functions and would rather see `FunctionParameterCount` extended to tackle closures as well.

File: CHANGELOG.md [hunk 2]
```diff
@@ -80,6 +80,11 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#1096](https://github.com/realm/SwiftLint/issues/1096)
 
+* Add `large_tuple` configurable rule that validates that tuples shouldn't
+  have more than 2 members.
```
Comment (jpsim) [inline CHANGELOG.md]: "shouldn't have too many members."

File: Source/SwiftLintFramework/Rules/LargeTupleRule.swift [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+public struct LargeTupleRule: ASTRule, ConfigurationProviderRule {
+
+    public var configuration = SeverityLevelsConfiguration(warning: 2, error: 3)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "large_tuple",
+        name: "Large Tuple",
+        description: "Tuples shouldn't have many members. Create a custom type instead.",
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/LargeTupleRule.swift]: "Tuples shouldn't have too many members"

File: Source/SwiftLintFramework/Rules/LargeTupleRule.swift [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            "func foo(bar: String) -> ↓(Int, Int, Int)\n",
+            "func foo(bar: String) -> ↓(Int, Int, Int) {}\n",
+            "func foo() throws -> ↓(Int, Int, Int)\n",
+            "func foo() throws -> ↓(Int, Int, Int) {}\n",
+            "func foo() throws -> ↓(Int, ↓(String, String, String), Int) {}\n"
+        ]
+    )
+
+    public func validateFile(_ file: File, kind: SwiftDeclarationKind,
+                             dictionary: [String : SourceKitRepresentable]) -> [StyleViolation] {
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/LargeTupleRule.swift]: Spacing: `dictionary: [String: SourceKitRepresentable])`

File: Source/SwiftLintFramework/Rules/LargeTupleRule.swift [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            let ranges = try? parenthesesRanges(type) else {
+                return []
+        }
+
+        var text = type.bridge()
+        var maxSize: Int?
+        for range in ranges {
+            let substring = text.substring(with: range)
+            let size = substring.components(separatedBy: ",").count
+            maxSize = max(size, maxSize ?? Int.min)
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/LargeTupleRule.swift]: Could use `?? .min` here.

**Comment by marcelofabri** [comment]
> Took me a while to validate all the little algorithms in here though 😬

I've actually used a playground to create the `parenthesesRanges` function first 😊

**Pull request merged**